### PR TITLE
Validator prompt acp77

### PR DESCRIPTION
**Pull request by sukantoraymond**

### CODE REVIEW

**Comment by meaghanfitzgerald** [comment]
How will the user generate the NodeID necessary for this input?

**Comment by meaghanfitzgerald** [comment]
How will the user generate the BLS info necessary for this input?

**Comment by meaghanfitzgerald** [comment]
What if someone wanted to preemptively generate their NodeIDs and BLS keys before spinning up their infrastructure? They should be able to generate these values locally so that they can send a successful `ConvertSubnetTx` _and then_ use these values to spin up nodes manually on cloud infrastructure instead of having to run multiple nodes locally and manage all the Ids and key pairs in various file locations.

**Review by felipemadero — COMMENTED** [review]
also, have we atm some functions or strutcs that we may want to set in SDK? this seems not to be the case pretty much, but probably we can mode at least node ID, BLS generation and validation, other stuff can not be moved until we better integrate SDK with subnet creation

File: cmd/blockchaincmd/create.go [hunk 1]
```diff
@@ -52,14 +50,16 @@ type CreateFlags struct {
 	validatorManagerMintOnly      bool
 	tokenMinterAddress            []string
 	validatorManagerController    []string
+	generateNodeID                bool
```
[the remark this answers is not in the document]
Comment (arturrez) [reply cmd/blockchaincmd/create.go:53]: not clear for me why we even need to generate NodeID

File: cmd/blockchaincmd/create.go [hunk 2]
```diff
@@ -218,6 +220,15 @@ func createBlockchainConfig(cmd *cobra.Command, args []string) error {
 		}
 	}
 
+	var bootstrapValidators []models.SubnetValidator
+	var err error
+	if bootstrapValidatorsJSONFilePath != "" {
+		bootstrapValidators, err = LoadBootstrapValidator(bootstrapValidatorsJSONFilePath)
```
Comment (arturrez) [inline cmd/blockchaincmd/create.go:226]: we need to check if fileExists here or inside of `LoadBootstrapValidator`

File: cmd/blockchaincmd/prompt_genesis_input.go [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+		return nil, cancelled, nil
+	}
+	return addr, false, nil
+}
+
+func promptProofOfPossession() (string, string, error) {
+	ux.Logger.PrintToUser("Next, we need the public key and proof of possession of the node's BLS")
+	ux.Logger.PrintToUser("Check https://docs.avax.network/api-reference/info-api#infogetnodeid for instructions on calling info.getNodeID API")
+	var err error
+	txt := "What is the public key of the node's BLS?"
```
Comment (arturrez) [inline cmd/blockchaincmd/prompt_genesis_input.go:76]: nit: better wording maybe `node's BLS public key`. pls ignore this comment :)

File: cmd/blockchaincmd/prompt_genesis_input.go [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+	publicKey, err := app.Prompt.CaptureValidatedString(txt, prompts.ValidateHexa)
+	if err != nil {
+		return "", "", err
+	}
+	txt = "What is the proof of possession of the node's BLS?"
```
Comment (arturrez) [inline cmd/blockchaincmd/prompt_genesis_input.go:81]: not clear for me what this `txt` means

File: cmd/blockchaincmd/prompt_genesis_input.go [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+	}
+	if createFlags.proofOfAuthority {
+		sidecar.ValidatorManagement = models.ValidatorManagementTypeFromString(proofOfAuthorityOption)
+		return nil
+	}
+	options := []string{proofOfAuthorityOption, proofOfStakeOption, explainOption}
+	var subnetTypeStr string
+	for {
+		option, err := app.Prompt.CaptureList(
+			"Which validator management protocol would you like to use in your blockchain?",
```
Comment (arturrez) [inline cmd/blockchaincmd/prompt_genesis_input.go:109]: nit: not sure if protocol is a correct word. maybe just `validator management` is good

File: pkg/constants/constants.go [hunk 6]
```diff
@@ -154,12 +154,14 @@ const (
...
 
-	TimeParseLayout             = "2006-01-02 15:04:05"
-	MinStakeWeight              = 1
-	DefaultStakeWeight          = 20
-	AVAXSymbol                  = "AVAX"
-	DefaultFujiStakeDuration    = "48h"
-	DefaultMainnetStakeDuration = "336h"
+	TimeParseLayout                  = "2006-01-02 15:04:05"
+	MinStakeWeight                   = 1
+	InitialBalanceBootstrapValidator = 1
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/constants/constants.go:159]: BootstrapValidatorMinBalance

File: pkg/constants/constants.go [hunk 7]
```diff
@@ -154,12 +154,14 @@ const (
...
+	DefaultWeightBootstrapValidator  = 1000000
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/constants/constants.go:160]: BootstrapValidatorDefaultWeight

File: pkg/networkoptions/network_options.go [hunk 8]
```diff
@@ -62,6 +62,31 @@ func networkOptionFromString(s string) NetworkOption {
 	return Undefined
 }
 
+func (n NetworkOption) ModelNetwork(devnetEndpoint string) (models.Network, error) {
```
Comment (arturrez) [inline pkg/networkoptions/network_options.go:65]: we should not name this as devnet. I would suggest naming it custom as currently we name it devnet but in the future we might have other networks to support.

File: pkg/networkoptions/network_options.go [hunk 9]
```diff
@@ -62,6 +62,31 @@ func networkOptionFromString(s string) NetworkOption {
...
+		if devnetEndpoint != "" {
+			infoClient := info.NewClient(devnetEndpoint)
+			ctx, cancel := utils.GetAPIContext()
+			defer cancel()
+			_, err := infoClient.GetNetworkID(ctx)
+			if err != nil {
+				return models.UndefinedNetwork, err
+			}
+		}
+		network = models.NewDevnetNetwork(devnetEndpoint, networkID)
```
Comment (arturrez) [inline pkg/networkoptions/network_options.go:81]: new custom network maybe

File: pkg/models/bootstrap_validator.go [hunk 10]
```diff
@@ -0,0 +1,17 @@
+// Copyright (C) 2022, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+package models
+
+type SubnetValidator struct {
+	NodeID string `json:"NodeID"`
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply pkg/models/bootstrap_validator.go:6]: not agree, but lets move on

File: cmd/blockchaincmd/prompt_genesis_input.go [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+		}
+		break
+	}
+	sidecar.ValidatorManagement = models.ValidatorManagementTypeFromString(subnetTypeStr)
+	return nil
+}
+
+// generateNewNodeAndBLS returns node id, bls public key and bls pop
+func generateNewNodeAndBLS() (string, string, string, error) {
+	certBytes, _, err := staking.NewCertAndKeyBytes()
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply cmd/blockchaincmd/prompt_genesis_input.go:122]: it is the same, but lets move on

File: cmd/blockchaincmd/prompt_genesis_input.go [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+			}
+			_, err := ids.NodeIDFromString(validatorJSON.NodeID)
+			if err != nil {
+				return fmt.Errorf("invalid node id %s", validatorJSON.NodeID)
+			}
+			if err = validateBLS(validatorJSON.BLSPublicKey, validatorJSON.BLSProofOfPossession); err != nil {
+				return err
+			}
+		}
+		if validatorJSON.Weight <= 0 {
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply cmd/blockchaincmd/prompt_genesis_input.go:232]: could you remove the < from <=

File: cmd/blockchaincmd/prompt_genesis_input.go [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+			return fmt.Errorf("bootstrap validator weight has to be greater than 0")
+		}
+		if validatorJSON.Balance <= 0 {
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply cmd/blockchaincmd/prompt_genesis_input.go:235]: same

File: pkg/prompts/validations.go [hunk 14]
```diff
@@ -129,6 +129,17 @@ func validateWeight(input string) error {
 	return nil
 }
 
+func validateBootstrapValidatorBalance(input string) error {
+	val, err := strconv.ParseUint(input, 10, 64)
+	if err != nil {
+		return err
+	}
+	if val <= 0 {
```
Comment (felipemadero) [inline pkg/prompts/validations.go:137]: val == 0 (it is unsigned)

File: pkg/models/bootstrap_validator.go [hunk 15]
```diff
@@ -0,0 +1,17 @@
+// Copyright (C) 2022, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+package models
+
+type SubnetValidator struct {
```
Comment (felipemadero) [inline pkg/models/bootstrap_validator.go:5]: BootstrapSubnetValidator ?

File: pkg/models/sidecar.go [hunk 16]
```diff
@@ -58,3 +59,14 @@ func (sc Sidecar) GetVMID() (string, error) {
 	}
 	return vmid, nil
 }
+
+func (sc Sidecar) NetworkDataIsEmpty(network string) bool {
+	if sc.Networks[network].SubnetID == ids.Empty &&
```
Comment (felipemadero) [inline pkg/models/sidecar.go:64]: you can just return this

Comment (felipemadero) [reply pkg/models/sidecar.go:64]: maybe you can do sc.Networks[network] == NetworkData{}

Comment (felipemadero) [reply pkg/models/sidecar.go:64]: or, _, b := sc.Networks[network], b (if you desire to check if the network is a key)

File: pkg/constants/constants.go [hunk 17]
```diff
@@ -151,12 +151,14 @@ const (
...
 
-	TimeParseLayout             = "2006-01-02 15:04:05"
-	MinStakeWeight              = 1
-	DefaultStakeWeight          = 20
-	AVAXSymbol                  = "AVAX"
-	DefaultFujiStakeDuration    = "48h"
-	DefaultMainnetStakeDuration = "336h"
+	TimeParseLayout                  = "2006-01-02 15:04:05"
+	MinStakeWeight                   = 1
+	InitialBalanceBootstrapValidator = 1
```
Comment (felipemadero) [inline pkg/constants/constants.go:156]: InitialBootstrapValidatorBalance?

Comment (felipemadero) [reply pkg/constants/constants.go:156]: add a comment for this two defaults so we remember the reason

File: cmd/blockchaincmd/prompt_owners.go [hunk 18]
```diff
@@ -289,3 +289,52 @@ func getThreshold(maxLen int) (uint32, error) {
 	}
 	return uint32(intTh), err
 }
+
+func getKeyForChangeOwner(previouslyUsedAddr string, network models.Network) (string, error) {
+	moreKeysPrompt := "Which key would you like to set as change owner for leftover AVAX if the node is removed from validator set?"
```
Comment (felipemadero) [inline cmd/blockchaincmd/prompt_owners.go:294]: can you use PromptAddress?

File: cmd/blockchaincmd/prompt_genesis_input.go [hunk 19]
```diff
@@ -0,0 +1,252 @@
...
+	"github.com/ava-labs/avalanche-cli/pkg/utils"
+	"github.com/ava-labs/avalanche-cli/pkg/ux"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/staking"
+	"github.com/ava-labs/avalanchego/utils/crypto/bls"
+	"github.com/ava-labs/avalanchego/utils/formatting"
+	"github.com/ava-labs/avalanchego/vms/platformvm/signer"
+)
+
+func getValidatorContractManagerAddr() ([]string, bool, error) {
```
Comment (felipemadero) [inline cmd/blockchaincmd/prompt_genesis_input.go:21]: this should return just string, not []string

Comment (felipemadero) [reply cmd/blockchaincmd/prompt_genesis_input.go:21]: getPoAValidatorContractOwnerAddr ?

Comment (felipemadero) [reply cmd/blockchaincmd/prompt_genesis_input.go:21]: keep PoA somewhere

Comment (felipemadero) [reply cmd/blockchaincmd/prompt_genesis_input.go:21]: if agree to return string, just use PromptAddress, not CaptureListDecision

File: cmd/blockchaincmd/prompt_genesis_input.go [hunk 20]
```diff
@@ -0,0 +1,252 @@
...
+		option, err := app.Prompt.CaptureList(
+			"Which validator management protocol would you like to use in your blockchain?",
+			options,
+		)
+		if err != nil {
+			return err
+		}
+		switch option {
+		case proofOfAuthorityOption:
+			subnetTypeStr = models.ProofOfAuthority
```
Comment (felipemadero) [inline cmd/blockchaincmd/prompt_genesis_input.go:108]: seems you can just assign to sidecar.ValidatorManagement

File: cmd/blockchaincmd/prompt_genesis_input.go [hunk 21]
```diff
@@ -0,0 +1,252 @@
...
+		case proofOfStakeOption:
+			subnetTypeStr = models.ProofOfStake
```
Comment (felipemadero) [inline cmd/blockchaincmd/prompt_genesis_input.go:110]: seems you can just assign to sidecar.ValidatorManagement

File: cmd/blockchaincmd/prompt_genesis_input.go [hunk 22]
```diff
@@ -0,0 +1,252 @@
...
+	explainOption := "Explain the difference"
+	if createFlags.proofOfStake {
+		sidecar.ValidatorManagement = models.ProofOfStake
+		return nil
+	}
+	if createFlags.proofOfAuthority {
+		sidecar.ValidatorManagement = models.ProofOfAuthority
+		return nil
+	}
+	options := []string{proofOfAuthorityOption, proofOfStakeOption, explainOption}
```
Comment (felipemadero) [inline cmd/blockchaincmd/prompt_genesis_input.go:96]: you can use the constants here

File: cmd/blockchaincmd/prompt_genesis_input.go [hunk 23]
```diff
@@ -0,0 +1,252 @@
...
+	for {
+		option, err := app.Prompt.CaptureList(
+			"Which validator management protocol would you like to use in your blockchain?",
+			options,
+		)
+		if err != nil {
+			return err
+		}
+		switch option {
+		case proofOfAuthorityOption:
```
Comment (felipemadero) [inline cmd/blockchaincmd/prompt_genesis_input.go:107]: you can use the constants here

File: cmd/blockchaincmd/prompt_genesis_input.go [hunk 24]
```diff
@@ -0,0 +1,252 @@
...
+			subnetTypeStr = models.ProofOfAuthority
+		case proofOfStakeOption:
+			subnetTypeStr = models.ProofOfStake
+		case explainOption:
```
Comment (felipemadero) [inline cmd/blockchaincmd/prompt_genesis_input.go:111]: you can even first check for this, and then just assing to sc field

**Review by felipemadero — COMMENTED** [review]
very close

**Review by felipemadero — APPROVED** [review]
LGTM, but probably artur should check his part

**Pull request merged**